pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -1184,168 +898,127 @@ protected IdealState ensureAllPartitionsConsuming(final TableConfig tableConfig,
-    try {
-      instanceAssignments = strategy.assign(Lists.newArrayList(newSegmentId), partitionAssignment);
-    } catch (InvalidConfigException e) {
-      _controllerMetrics.addMeteredTableValue(idealState.getResourceName(),
-          ControllerMeter.CONTROLLER_REALTIME_TABLE_SEGMENT_ASSIGNMENT_ERROR, 1L);
-      throw new IllegalStateException("Caught exception when updating ideal state on segment completion", e);
+  private int getNumPartitionsFromIdealState(IdealState idealState) {
+    int numPartitions = 0;
+    for (String segmentName : idealState.getRecord().getMapFields().keySet()) {
+      if (LLCSegmentName.isLowLevelConsumerSegmentName(segmentName)) {
+        numPartitions = Math.max(numPartitions, new LLCSegmentName(segmentName).getPartitionId() + 1);
+      }
     }
+    return numPartitions;
+  }
 
-    List<String> newSegmentInstances = instanceAssignments.get(newSegmentId);
-    Set<String> currentSegmentInstances = idealState.getInstanceSet(currentSegmentId);
-    for (String instance : currentSegmentInstances) {
-      idealState.setPartitionState(currentSegmentId, instance,
-          PinotHelixSegmentOnlineOfflineStateModelGenerator.ONLINE_STATE);
+  private int getNumReplicas(TableConfig tableConfig, InstancePartitions instancePartitions) {
+    if (instancePartitions.getNumReplicaGroups() == 1) {
+      // Non-replica-group based
+      return tableConfig.getValidationConfig().getReplicasPerPartitionNumber();
+    } else {
+      // Replica-group based
+      return instancePartitions.getNumReplicaGroups();
     }
+  }
 
-    // The {@link RealtimeSegmentValidationManager} will fix metadata and ideal state after {@link MAX_SEGMENT_COMPLETION_TIME_MILLIS} of inactivity on the committing segment
-    // If the ideal state update during completion took longer than {@link MAX_SEGMENT_COMPLETION_TIME_MILLIS},the update could already have been done by the fixer thread .
-    // We do not want to overwrite the ideal state. It is possible that the new segment created by the fixer has already progressed to ONLINE.
-    // If we let the below update happen, we will be bringing an ONLINE segment back to CONSUMING, and end up with 2 CONSUMING segments for the partition
-    Map<String, String> stateMap = idealState.getInstanceStateMap(newSegmentId);
-    if (stateMap == null) {
-      for (String instance : newSegmentInstances) {
-        idealState.setPartitionState(newSegmentId, instance,
-            PinotHelixSegmentOnlineOfflineStateModelGenerator.CONSUMING_STATE);
-      }
+  private int getMaxNumPartitionsPerInstance(InstancePartitions instancePartitions, int numPartitions,
+      int numReplicas) {
+    if (instancePartitions.getNumReplicaGroups() == 1) {
+      // Non-replica-group based assignment:
+      // Uniformly spray the partitions and replicas across the instances.
+      // E.g. (6 instances, 3 partitions, 4 replicas)
+      // "0_0": [i0,  i1,  i2,  i3,  i4,  i5  ]
```

Comment:
typo in this example -> p0r0 p0r1 p0r2 `p0r3`

Reply:
Good catch